Logging fixes and enhancements

bart-vmware opened the pull request:
## Description

Logging
- Reuse existing console formatters.
- Fixed: scopes do not work when enabled through configuration.
- Fixed: turn off colors when on CloudFoundry.

To clarify: before this PR, `DynamicConsoleLoggerProvider` would instantiate `ConsoleLoggerProvider`, passing in options but not formatters. This resulted in the default formatters still being instantiated under the covers, but ignoring configuration from appsettings.json. The effect is that it wasn't possible to turn on scopes or turn off console colors. The fix is to obtain a `ConsoleLoggerProvider` from the D/I container instead, which calls the proper constructor that includes the `formatters` parameter.

The logic to control and test colors used to be wrapped in a `#if ... #endif` because `ConsoleLoggerOptions.DisableColors` was deprecated after .NET Core 3.1 in favor of `SimpleConsoleFormatterOptions.ColorBehavior`. When Steeltoe dropped support for .NET Core 3.1, the related code was deleted. This PR restores that functionality.

## Quality checklist

- [x] Your code complies with our [Coding Style](https://github.com/SteeltoeOSS/.github/blob/main/contributing-docs/contributing-code-style.md).
- [x] You've updated unit and/or integration tests for your change, where applicable.
- [ ] You've updated documentation for your change, where applicable.
      If your change affects other repositories, such as [Documentation](https://github.com/SteeltoeOSS/Documentation), [Samples](https://github.com/SteeltoeOSS/Samples) and/or [MainSite](https://github.com/SteeltoeOSS/MainSite), add linked PRs here.
- [ ] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.
- [ ] You've added required license files and/or file headers (explaining where the code came from with proper attribution), where code is copied from StackOverflow, a blog, or OSS.

TimHess commented on src/Logging/src/Abstractions/DynamicLoggerProviderBase.cs:
```diff
@@ -333,4 +333,12 @@ private IEnumerable<string> GetKeyPrefixes(string name)
         return null;
     }
+
+    public void SetScopeProvider(IExternalScopeProvider scopeProvider)
```
How or when is this used?

bart-vmware replied:
For scopes to work, `ConsoleLogger` needs a scope provider [here](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs#L79). It is then subsequently passed on to the formatter [here](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging.Console/src/ConsoleLogger.cs#L56), which renders the scope values. `LoggerFactory` creates the scope provider and exposes it to logger providers  [here](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging/src/LoggerFactory.cs#L197-L202).

TimHess replied:
Oh! I missed the connection with `ISupportExternalScope`

TimHess commented on src/Management/test/Endpoint.Test/Loggers/TestLogProvider.cs:
```diff
@@ -35,4 +35,8 @@ public void SetLogLevel(string category, LogLevel? level)
         Level = level ?? LogLevel.None;
     }
+
+    public void SetScopeProvider(IExternalScopeProvider scopeProvider)
```
Is this supposed to be empty?

hananiel commented on src/Logging/src/Abstractions/DynamicLoggerProviderBase.cs:
```diff
@@ -333,4 +333,12 @@ private IEnumerable<string> GetKeyPrefixes(string name)
 
         return null;
     }
+
```
```suggestion
/// <inheritdoc cref="ISupportExternalScope"/>
```

hananiel replied:
Since it's public it helps to have some docs even though its only ever called by framework.

bart-vmware replied:
Agreed. Turns out just adding it without `cref` works as well:

![image](https://user-images.githubusercontent.com/104792814/206719768-d3e89be6-a837-40ac-895a-5618551215c8.png)

hananiel commented on src/Logging/src/Abstractions/MessageProcessingLogger.cs:
```diff
@@ -46,31 +46,32 @@ public bool IsEnabled(LogLevel logLevel)
             }
         }
 
-        if (!string.IsNullOrEmpty(message) || exception != null)
```
This  null/empty message or exception was removed- was this intentional?

bart-vmware replied:
Yes. The logic has moved into the formatters, for example: https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Logging.Console/src/SimpleConsoleFormatter.cs#L50-L53

hananiel commented on src/Logging/src/DynamicLogger/ConsoleLoggerOptionsSetup.cs:
```diff
@@ -1,17 +0,0 @@
-
-namespace Steeltoe.Logging.DynamicLogger;
-
-internal sealed class ConsoleLoggerOptionsSetup : ConfigureFromConfigurationOptions<ConsoleLoggerOptions>
```
I don't know if this impacts how we are using this but not all of `ConsoleLoggerOptions`[link](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.console.consoleloggeroptions?view=dotnet-plat-ext-7.0) has been deprecated - just most of it. Do the remaining two have a bearing on how our dynamic logger is  configured?

bart-vmware replied:
[MaxQueueLength](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.console.consoleloggeroptions.maxqueuelength?view=dotnet-plat-ext-7.0#microsoft-extensions-logging-console-consoleloggeroptions-maxqueuelength) and [QueueFullMode](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.console.consoleloggeroptions.queuefullmode?view=dotnet-plat-ext-7.0#microsoft-extensions-logging-console-consoleloggeroptions-queuefullmode) were added in .NET 7, [LogToStandardErrorThreshold](https://learn.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.console.consoleloggeroptions.logtostandarderrorthreshold?view=dotnet-plat-ext-7.0#microsoft-extensions-logging-console-consoleloggeroptions-logtostandarderrorthreshold) was introduced in .NET Core 3.

These settings are not relevant for dynamic logging, however they are consumed by its wrapped logger. In that case, I don't think we need to do anything to have the built-in console logger consume these settings.

bart-vmware replied:
I've manually tested that setting `Logging:Console:LogToStandardErrorThreshold = "Information"` in appsettings.json works when dynamic logging is used. Similarly to scopes, this is pretty hard to unit-test.

bart-vmware replied:
Added a test that uses the obsolete boolean `DisableColors`, so this should be covered now.

hananiel commented on src/Logging/src/DynamicLogger/DynamicLoggingBuilder.cs:
```diff
@@ -24,34 +24,69 @@ public static ILoggingBuilder AddDynamicConsole(this ILoggingBuilder builder)
+        }
+
+        // Yet we need an instance from the container to construct DynamicConsoleLoggerProvider, so register without interface.
+        services.AddSingleton<ConsoleLoggerProvider>();
```
👍

hananiel commented on src/Logging/src/DynamicLogger/DynamicLoggingBuilder.cs:
```diff
@@ -24,34 +24,69 @@ public static ILoggingBuilder AddDynamicConsole(this ILoggingBuilder builder)
-
             builder.Services.AddSingleton(p => p.GetServices<ILoggerProvider>().OfType<IDynamicLoggerProvider>().SingleOrDefault());
+
+            DisableConsoleColorsOnCloudFoundry(builder);
```
Are we sure it is actually still needed (for example see: https://github.com/dotnet/runtime/issues/37421) b. Is this needed in K8s/TAP and if so it might be named differently.

TimHess replied:
I think @dtillman mentioned on a call today that it is still needed and is also needed on TAP, so we should probably rename this and switch to `Platform.IsCloudHosted`

hananiel reviewed (changes requested):
Do  we need some tests to actually verify scoped logging is working? For example all our `Logging:IncludeScopes` is false and I believe that is no longer the correct path for that configuration.

bart-vmware commented on src/Common/test/Common.TestResources/TestHelpers.cs:
```diff
@@ -86,7 +87,7 @@ public static ILoggerFactory GetLoggerFactory()
     {
         IServiceCollection serviceCollection = new ServiceCollection();
         serviceCollection.AddLogging(builder => builder.SetMinimumLevel(LogLevel.Trace));
-        serviceCollection.AddLogging(builder => builder.AddConsole());
```
Thanks, I've removed the helper.

bart-vmware commented:
> Do we need some tests to actually verify scoped logging is working? For example all our Logging:IncludeScopes is false and I believe that is no longer the correct path for that configuration.

You're right, `Logging:IncludeScopes` does not work. Although obsolete, `Logging:Console:IncludeScopes` still works.

The modern way is to use `Logging:Console:FormatterOptions:IncludeScopes`, BUT that only works when _also_ specifying a formatter explicitly by either:
- adding `Logging:Console:FormatterName="Simple|Systemd|Json"` in configuration
- calling `services.AddLogging(options => options.Add[Simple|Systemd|Json]Console(formatterOptions => formatterOptions.IncludeScopes = true))` instead of the default `.AddConsole()`

I've added configuration-based tests for both, and updated all other tests to use the obsolete `Logging:Console:IncludeScopes` instead of `Logging:IncludeScopes`. That way, those tests keep working whether any of the three formatters is specified or none at all.
